flocker/ca/_ca.py

```diff
@@ -0,0 +1,225 @@
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+Low-level logic for a certificate authority.
+
+Uses ECDSA + SHA256? Or if that turns out to be too difficult RSA 4096-bit
++ SHA 256.
+"""
+
+import os
+
+from OpenSSL import crypto
+from pyrsistent import PRecord, field
+from twisted.internet.ssl import DistinguishedName, KeyPair, Certificate
+
+
+EXPIRY_20_YEARS = 60 * 60 * 24 * 365 * 20
+
+
+class CertificateAlreadyExistsError(Exception):
+    """
+    Error raised when a certificate file already exists.
+    """
+
+
+class KeyAlreadyExistsError(Exception):
+    """
+    Error raised when a keypair file already exists.
+    """
+
+
+class PathError(Exception):
+    """
+    Error raised when the directory for certificate files does not exist.
+    """
+
+
+class FlockerKeyPair(object):
+    """
+    KeyPair with added functionality for comparison and signing a request
+    object with additional extensions for generating a self-signed CA.
+    """
+    def __init__(self, keypair):
+        self.keypair = keypair
+
+    def __eq__(self, other):
+        if isinstance(other, FlockerKeyPair):
+            return self.keypair.dump() == other.keypair.dump()
+        return False
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def selfSignedCACertificate(self, dn, request, serial, expiry, digest):
+        """
+        Sign a CertificateRequest with extensions for use as a CA certificate.
+
+        This code based on ``twisted.internet.ssl.KeyPair.signRequestObject``
+
+        :param DistinguishedName dn: The ``DistinguishedName`` for the
+            certificate.
+
+        :param CertificateRequest request: The signing request object.
+
+        :param int serial: The certificate serial number.
+
+        :param int expiry: Number of seconds from now until this certificate
+            should expire.
+
+        :param str digest: The digest algorithm to use.
+        """
+        req = request.original
+        cert = crypto.X509()
+        dn._copyInto(cert.get_issuer())
+        cert.set_subject(req.get_subject())
+        cert.set_pubkey(req.get_pubkey())
+        cert.gmtime_adj_notBefore(0)
+        cert.gmtime_adj_notAfter(expiry)
+        cert.set_serial_number(serial)
+        cert.add_extensions([
+            crypto.X509Extension("basicConstraints", True,
+                                 "CA:TRUE, pathlen:0"),
+            crypto.X509Extension("keyUsage", True,
+                                 "keyCertSign, cRLSign"),
+            crypto.X509Extension("subjectKeyIdentifier", False, "hash",
+                                 subject=cert),
+        ])
+        cert.add_extensions([
+            crypto.X509Extension(
+                "authorityKeyIdentifier", False,
+                "keyid:always", issuer=cert
+            )
+        ])
+        cert.sign(self.keypair.original, digest)
+        return Certificate(cert)
+
+
+def generate_keypair():
+    """
+    Create a new key pair.
```

Comment:
Document the return type and specifically that it will be a RSA 4096bit key.

And I guess this answers the comment in the module docstring about ECDSA keys maybe being too difficult.